numba/ir_utils.py

```diff
@@ -2020,3 +2020,62 @@ def check_and_legalize_ir(func_ir):
         msg +=  func_ir.diff_str(orig_ir)
         warnings.warn(NumbaWarning(msg, loc=func_ir.loc))
 
+
+def convert_code_obj_to_function(code_obj, caller_ir):
+    """
+    Converts a code object from a `make_function.code` attr in the IR into a
+    python function, caller_ir is the FunctionIR of the caller and is used for
+    the resolution of freevars.
+    """
+    fcode = code_obj.code
+    nfree = len(fcode.co_freevars)
+
+    # try and resolve freevars if they are consts in the caller's IR
+    # these can be baked into the new function
+    freevars = []
+    for x in fcode.co_freevars:
+        freevar_def = caller_ir.get_definition(x)
+        if isinstance(freevar_def, ir.Const):
+            freevars.append(freevar_def.value)
+        else:
+            msg = ("Cannot capture a non-constant value for variable '%s' in a "
+                   "function that will escape." % x)
+            raise TypingError(msg, loc=code_obj.loc)
+
+    func_env = "\n".join(["  c_%d = %s" % (i, x) for i, x in enumerate(freevars)])
+    func_clo = ",".join(["c_%d" % i for i in range(nfree)])
+    func_arg = ",".join(["x_%d" % i for i in range(fcode.co_argcount)])
+    func_text = "def g():\n%s\n  def f(%s):\n    return (%s)\n  return f" % (
+        func_env, func_arg, func_clo)
+    loc = {}
+    # globals are the same as those in the caller
+    glbls = caller_ir.func_id.func.__globals__
+    exec_(func_text, glbls, loc)
```

Comment:
@ehsantn I think this might in part fix up the colloquial "code objects don't handle freevars properly" issue?

Reply:
Yes, looks like this might work! It probably needs careful testing to be sure. BTW, the free var constant requirement is not strictly necessary in all contexts I think, since the IR can just be inlined in which case the variable becomes available.

Reply:
That may indeed work. However, use case for the `make_function`s here are that they have to be able to escape as a fully fledged jit function so I think freevars have to be constants cf. globals as const for a module level function. I don't recall, is there a simple case recorded somewhere where freevars are not working properly?

This patch needs more effort and splitting out anyway to have closures escaping as functions as a separate concern to supporting `map` and friends. I think the first part would be a good place to re-engineer the above to do any fix-ups for other potential problems.

Reply:
I don't remember if we have issues opened about freevars either.